Updated balancing documentation (#1226)
Haivision/srt: docs/socket-groups.md

## 2. Backup

> This solution is more complicated and more challenging for the settings,

Reviewer: ```suggestion
The settings for this grout type are more complicated and more challenging.
```

Author: What I actually meant is that it may be challenging to define proper settings by the user in accordance to network conditions and desired latency, and "complicated" is the overall solution of the backup groups. I rephrased it a little bit differently.

Reviewer: "The configuration of Backup groups is somewhat more complicated than with the other group types. In particular, it may be challenging to arrive at the optimal settings for a a given set of network conditions and desired latency."

> and in contradiction to Broadcast group, it costs some penalties. It has
> also advantages: the overhead for redundancy, which in case of broadcast
> groups is 100% per every next link, in this case it is tried to be kept at
> negligible minimum.

Reviewer: ```suggestion
Unlike with broadcast groups, there are some penalties, but there are
also advantages. Whereas the overhead for redundancy in the case of broadcast
groups is 100% per every next link, this is usually kept at
a negligible minimum for backup groups.
```

> In this group, in a normal situation, only one link out of member links is used
> for transmission, while others are just kept alive (the keepalive message
> is sent over them usually once per 1 second). Other links may start being used
> when there's happening an event of "disturbance" on a link, which makes it
> considered "unstable". This term is introduced beside "broken" because SRT
> normally uses 5 seconds to be sure that the link is broken, and this is way too
> much to be used as a latency penalty, if you still want to have a relatively
> low latency.

Reviewer: ```suggestion
Under normal circumstances, only one of the member links in this group is used
for transmission, while the others are just kept alive (a keepalive message
is sent over these links usually once per 1 second). Other links may start being used
when a "disturbance" event occurs on a link, whereupon the link is
considered to be "unstable". There is a distinction to be made here between "unstable" 
and "broken". SRT normally waits 5 seconds to be sure that a link is broken, which is too
much of a penalty if you want to have a relatively low latency. SRT can react more quickly 
to an "unstable" link condition.
```

Author: I kept this, but added some text at the end.

> for SRT to determie the link broken - either by getting the link broken by
> itself, or by closing the link when it's remaining unstable too long time.

Reviewer: ```suggestion
for SRT to determine the link broken - either by breaking the link by
itself, or by closing the link when it has been unstable too long.
```

The overhead here may then come from two sources:
- the keepalive control packets for inactive links (negligible)
- sending data over an unstable link for the resolution time

> This means that the overhead is usually varrying, depending on how often

Reviewer: ```suggestion
This means that the overhead is usually varying, depending on how often
```

> This means that the overhead is usually varrying, depending on how often
> it happens that a link becomes unstable. It is then very important that
> you tweak the value of stability timeout properly and cautiosly. If this
> time is too small, your link might be too often too eagerly qualified as
> unstable and with every case of unstable link it costs you temporary
> overhead. Note that a single activation of a backup link costs you even
> more than 100% overhead for the short time (up to 5 seconds), as usually some
> packets that have been already sent over this unstable link, but weren't
> acknowledged yet, will be sent again over the newly activated link. As
> this is usually tolerable minimum overhead if it was due to a broken link,
> you may experience link switching much more often and uselessly, if your
> stability timeout is too small.

Reviewer: ```suggestion
This means that the overhead is usually varying, depending on how often 
a link becomes unstable. It is then very important that you tweak the value 
of the stability timeout properly and cautiously. If this value is too small, 
your link might be too eagerly qualified as unstable, costing temporary 
overhead. Note that a single activation of a backup link costs more than 
100% overhead for up to 5 seconds, since any packets already sent over 
this unstable link (but not acknowledged) will be sent again over the newly 
activated link. This is usually a tolerable minimum overhead when due to a 
broken link. But you may experience unnecessary link switching much more 
often if your stability timeout is too small.
```

This mode allows also to set link priorities, through the `weight`
parameter - the lower, the more preferred. This priority decides mainly, which
link is "best" and which is selected to take over transmission over a broken
link before others, as well as which links should remain active should multiple
links be stable at a time.  If you don't specify priorities, the second
connected link need not take over sending, although as this is resolved through
sorting, then whichever link out of those with the same priority would take

> over when all links are stable is undefined.

Reviewer: ```suggestion
This mode allows you to set link priorities using the `weight` parameter 
(a lower value is preferred). The priority determines which link is "best" and 
will be selected to resume transmission over a broken link. When there are 
multiple stable links, the priority also determines which should remain active. 
 If you don't specify priorities, then which of the connected links will take over 
 sending is undefined, although this is resolved through sorting.
```

Author: Well, I misspelled something again. What I meant was that the lower values of the `weight` parameter gives the link higher priority rank - a link with weight=0 is more preferred than a link with weight=1.

Reviewer: Fixed on Line 245

that packets lost on the broken link can be resent over the others,
but no such mechanism has been provided for balancing group.

Please also keep in mind that there's a rule for all groups - one member
link established is enough for a group to be connection established. This
might be not always wanted in case of balancing groups, so your application
might need to do additional check and monitor how many links and which
ones are currently active. For example, if you need 3 links to balance
the load and having only 2 of them would be too little to withstand the
whole transmission, you should wait with connecting until all 3 links
are established. Important thing here is also that one of the links might
be at some moment not possible to be established, or one of the links
can get broken during transmission.

> 

Reviewer: ```suggestion
Please also keep in mind that this rule applies to all groups: *establishing 
one member link is enough for a group connection to be established.* But 
when balancing groups, this might not always be wanted. In such cases, 
your application might need to monitor how many links are established, and 
which ones are currently active. For example, if you need a minimum of 3 links 
to balance the load, you should delay connecting until all 3 links are established. 
It is also important to take into consideration that it might not be possible to establish 
one of the links, or one of the links might get broken during transmission.

```

Author: Partially distributed in other parts; text to be kept copied from here.

Author: This has been distributed more in other places, so the text was mainly derived from this, but also changed.

As there could be various ways as to how to implement balancing
algorithm, there's a framework provided to implement various methods,

> and two algorithms are currently provided:

Author: I'd remove this "by default" - the situation is that the framework is prepared, but - unlike e.g. packetfilter - this is not exposed to the user. New algorithms can only be added as a part of the SRT source code; the framework just makes the future upgrades with new algorithms easier.

> 1. `fixed`. In this algorithm you specify the share that is given to
> particular link manually through the `weight` parameter. The weight
> value in case of this algorithm in balancing groups defines how much of
> a share is given to this link. Important here is that you can easily
> think of the weight values as a percentage of load burden for particular
> link - however in reality the share of the load is calculated as a
> percentage that particular link's weight comprises among the sum of all
> weight values. Additionally, a value of 0 is special and is translated
> into a weight that would make it equal to an average of all weights.
> Be careful here with the non-established and broken links. For example,
> if you have 3 links with weight 10, 20 and 30, it results in a load
> balance of 16.6%, 33.3% and 50% respectively. However if the second link
> gets broken, there are then 2 links with 10 and 30, which results in
> load balance of 25% and 75% respectively. Keep in mind that it's up
> to the application to keep the minimum links allowed and break the
> group link that is unable to withstand the load.

Reviewer: ```suggestion
1. `fixed`. In this algorithm you manually specify the share of a load that is given to
a particular link through the `weight` parameter. In Balancing groups, the weight
value defines how much of a share of the load is given to a link. That share is 
calculated as a percentage of the sum of all weight values. For example, if you have 
three links with weights of 10, 20 and 30, this results in a load balance of 16.6%, 33.3% 
and 50% respectively. 

Note that a value of 0 is translated into a weight that would make it equal to an average 
of all non-zero weights. For example, if you have weights A, B and C of 0, 20, and 30 
respectively, then A = (B+C)/2 = 25 (any other links in the group with weight = 0 would 
also inherit the same calculated weight). If all weights are equal to zero, then they are 
all nominally set to a value of 1.

Be careful here with non-established and broken links. In the example above, if the 
second link gets broken, there are then two links with weights of 10 and 30, which 
results in load balance values of 25% and 75% respectively. Keep in mind that it's up
to the application to monitor for the minimum number of active links needed, and to 
break any group link that is unable to withstand the load.
```

> 2. `window` (default). This algorithm is performing cyclic measurement of the

Reviewer: ```suggestion
2. `window` (default). This algorithm performs cyclic measurement of the
```

> minimum flight window and this way determines the "cost of sending"
> of a packet over particular link. The link is then "paid" for sending
> a packet appropriate "price", which is collected in the link's "pocket".
> To send the next packet the link with lowest state of the "pocket" is
> selected. The "cost of sending" measurement is being repeated once per
> a time with a distance of 16 packets on each link.

Reviewer: ```suggestion
minimum flight window and this way determines the "cost of sending"
of a packet over a particular link. The link is then "paid" for sending
a packet with an appropriate "price", which is collected in the link's "purse".
The link with lowest amount in its "purse" is selected to send the next packet. 
The "cost of sending" measurement is being repeated once per
a time at an interval of 16 packets on each link.
```

Author: Rephrased to be more "readable". My original idea was a little bit too "metaphoric".

> There are possible also other methods and algorithms in the future.

Reviewer: ```suggestion
The addition of other methods and algorithms is permitted.
```

Author: I'd still keep "possible in future". What I meant here is that future changes in the SRT source code may be done to add new algorithms.

Reviewer: Adding temporal qualifiers like "currently" or "in the latest version" or "in the future" is usually not good practice. As time goes by, the context is lost. I don't think it's necessary to make promises here.

that packets lost on the broken link can be resent over the others,
but no such mechanism has been provided for balancing group.

Please also keep in mind that the group is considered connected when
it contains at least one connected member link. This means also that the
group becomes ready for transmission after connecting the first link,
as well as it remains ready even if some member links get broken.

If the application wants to make sure that a transmission is balanced between
links (where only together can they maintain the bandwidth capacity required
for a signal), it must make sure that all "required" links are established by
monitoring the group data. For example, if you need a minimum of 3 links to
balance the load, you should delay starting the transmission until all 3 links
are established (that is, all of them report "idle" state), and also stop it in
case when a broken link caused that the others do not cover the required
capacity.

> Please also keep in mind that there's a rule for all groups - one member

Author: This text should be removed. Statements are covered already.
